File: prob_QCQP/run_qcqp.m
% Example script to run the TI-DSFO algorithm to solve the TRO problem

% Author: Cem Musluoglu, KU Leuven, Department of Electrical Engineering
% (ESAT), STADIUS Center for Dynamical Systems, Signal Processing and Data
% Analytics
% Correspondence: cemates.musluoglu@esat.kuleuven.be

clear all
close all

addpath('../DSFO/');

% Number of Monte-Carlo runs
mc_runs=5;

% Number of nodes
nbnodes=10;
% Number of channels per node
nbsensors_vec=5*ones(1,nbnodes);
% Number of channels in total
nbsensors=sum(nbsensors_vec);

% Number of samples of the signals
nbsamples=10000;

% Number of filters of X
Q=3;

norm_error=cell(mc_runs,1);
n_runs=1;

while n_runs<=mc_runs
    
    % Create the data
    [Y,B,alpha,c,d]=create_data(nbsensors,nbsamples,Q);
    
    % Structure related to parameters of the problem
    Y_cell{1}=Y;
    B_cell{1}=B;
    B_cell{2}=c;
    Gamma_cell{1}=eye(nbsensors);
    Glob_Const_cell{1}=alpha;
    Glob_Const_cell{2}=d;
    
    data=struct;
    data.Y_cell=Y_cell;
    data.B_cell=B_cell;
    data.Gamma_cell=Gamma_cell;
    data.Glob_Const_cell=Glob_Const_cell;

    % Structure related to parameters of the problem
    prob_params=struct;
    prob_params.nbsensors=nbsensors;
    prob_params.Q=Q;
    prob_params.nbnodes=nbnodes;
    prob_params.nbsensors_vec=nbsensors_vec;
    prob_params.nbsamples=nbsamples;
    
    % Estimate filter using the centralized algorithm
    [X_star,f_star]=qcqp_solver(data,prob_params);
    prob_params.X_star=X_star;
    % Compute the distance to X^* if equal to 1 
    prob_params.compare_opt=1;
    % Show a dynamic plot if equal to 1
    prob_params.plot_dynamic=0;

    % Structure related to stopping conditions. The last to be
    % achieved stops the algorihtm (To choose only one condition, set the
    % other to a negative value).
    conv=struct;
    conv.tol_f=-1;
    conv.nbiter=200;
    
    % Create adjacency matrix (hollow matrix) of a random graph
    adj=randi(2,nbnodes,nbnodes)-1;
    graph_adj=triu(adj,1)+tril(adj',-1);
    prob_params.graph_adj=graph_adj;
    
    try
        [X_est,f_seq,norm_diff,norm_err]=dsfo(data,prob_params,...
                                conv,@qcqp_eval,@qcqp_solver,[]);
        norm_error{n_runs}=norm_err;
        n_runs=n_runs+1;
    catch
        warning('Infeasible')
    end


end

%%
% Plot the normalized error

x_int=[1:conv.nbiter];
q_5=quantile(cell2mat(norm_error),0.5);
q_25=quantile(cell2mat(norm_error),0.25);
q_75=quantile(cell2mat(norm_error),0.75);
loglog(q_5,'b','LineWidth',2);

hold on
fill([x_int,fliplr(x_int)],[q_5,fliplr(q_75)],'b','FaceAlpha','0.2','LineStyle','none')
fill([x_int,fliplr(x_int)],[q_5,fliplr(q_25)],'b','FaceAlpha','0.2','LineStyle','none')
xlim([1,inf])
ylim([1e-6,inf])

ax=gca;
ax.FontSize=14;
xlabel('Iterations','Interpreter','latex','Fontsize',20)
ylabel('$\frac{||X^{i}-X^*||_2^F}{||X^*||_F^2}$','Interpreter','latex','Fontsize',20)
grid on

%%

function [Y,B,alpha,c,d]=create_data(nbsensors,nbsamples,Q)

    rng('shuffle');
    Y=randn(nbsensors,nbsamples);
    Y=Y-mean(Y,2);
    Y=Y./var(Y')';
    Ryy=1/nbsamples*conj(Y*Y');
    Ryy=make_sym(Ryy);
    B=randn(nbsensors,Q);
    c=randn(nbsensors,1);
    d=randn(Q,1);
    w=(B'*inv(Ryy)'*c-d)./(c'*inv(Ryy)*c);
    X=inv(Ryy)*(B-c*w');
    
    toss=randi(2)-1;
    
    if toss==0
        alpha=randn(1,1);
        alpha=alpha^2;
    % Enforce the solution to be strictly satisfying the inequality
    % constraint
    else
        alpha=randn(1,1);
        alpha=alpha^2;
        alpha=sqrt(norm(X,'fro')^2+alpha^2);
    end

    % If problem is infeasible, recreate the parameters
    while(alpha^2<norm(d)^2/norm(c)^2)
        c=randn(nbsensors,1);
        d=randn(1,1);
        w=(B'*inv(Ryy)'*c-d)/(c'*inv(Ryy)*c);
        X=inv(Ryy)*(B-c*w');
        toss=randi(2)-1;
        if toss==0
            alpha=randn(1,1);
            alpha=alpha^2;
        else
            alpha=randn(1,1);
            alpha=alpha^2;
            alpha=sqrt(norm(X,'fro')^2+alpha^2);
        end
    end

end

File: Python_code/prob_QCQP/run_qcqp.py
import numpy as np
import sys
import matplotlib as mpl
# Choose plot backend.
# mpl.use('macosx')
# mpl.use('Qt5Agg')
# mpl.use('TkAgg')
mpl.use('Agg')
import matplotlib.pyplot as plt
from time import sleep

sys.path.append('../dasf_toolbox/')
import qcqp_functions as qcqp
from dasf_toolbox import dasf
from dasf_toolbox import dasf_block

# Number of Monte-Carlo runs.
mc_runs = 5

# Number of nodes.
nbnodes = 10
# Number of channels per node.
nbsensors_vec = 5 * np.ones(nbnodes)
nbsensors_vec = nbsensors_vec.astype(int)
# Number of channels in total.
nbsensors = np.sum(nbsensors_vec)

# Number of samples of the signals.
nbsamples = 10000

# Number of filters of X.
Q = 3

norm_error = []
n_runs = 0

rng = np.random.default_rng()

while n_runs < mc_runs:
    # Create the data.
    Y, B, alpha, c, d = qcqp.create_data(nbsensors, nbsamples, Q)

    # Dictionary related to the data of the problem.
    Y_list = [Y]
    B_list = [B, c]
    Gamma_list = [np.identity(nbsensors)]
    Glob_Const_list = [alpha, d]

    data = {'Y_list': Y_list, 'B_list': B_list,
            'Gamma_list': Gamma_list, 'Glob_Const_list': Glob_Const_list}

    # Dictionary related to parameters of the problem.
    prob_params = {'nbnodes': nbnodes, 'nbsensors_vec': nbsensors_vec,
                   'nbsensors': nbsensors, 'Q': Q, 'nbsamples': nbsamples}

    # Create adjacency matrix (hollow matrix) of a random graph.
    graph_adj = rng.integers(0, 1, size=(nbnodes, nbnodes), endpoint=True)
    graph_adj = np.triu(graph_adj, 1) + np.tril(graph_adj.T, -1)
    prob_params['graph_adj'] = graph_adj

    # Random updating order.
    update_path = rng.permutation(range(nbnodes))
    prob_params['update_path'] = update_path

    # Estimate filter using the centralized algorithm.
    X_star = qcqp.qcqp_solver(prob_params, data)
    f_star = qcqp.qcqp_eval(X_star, data)

    prob_params['X_star'] = X_star
    # Compute the distance to X_star if "True".
    prob_params['compare_opt'] = True
    # Show a dynamic plot if "True".
    prob_params['plot_dynamic'] = False

    # Dictionary related to stopping conditions. We fix the number of iterations the DASF algorithm will perform to 200.
    nbiter = 200
    conv = {'nbiter': nbiter}

    try:
        # Solve the QCQP in a distributed way using the DASF framework.
        X_est, norm_diff, norm_err, f_seq = dasf(prob_params, data, qcqp.qcqp_solver,
                                                 conv, prob_eval=qcqp.qcqp_eval, prob_select_sol=None)

        norm_error.append(norm_err)
        n_runs = n_runs + 1

        sys.stdout.write('\r')
        j = (k + 1) / mc_runs
        sys.stdout.write("[%-20s] %d%%" % ('='*int(20*j), 100*j))
        sys.stdout.flush()
        sleep(0.25)
    except:
        print("Infeasible")

sys.stdout.write('\n')


# Plot the normalized error.
q5 = np.quantile(norm_error, 0.5, axis=0)
q25 = np.quantile(norm_error, 0.25, axis=0)
q75 = np.quantile(norm_error, 0.75, axis=0)
iterations = np.arange(1, nbiter + 1)

fig = plt.figure()
ax = fig.add_subplot(1, 1, 1)
ax.loglog(iterations, q5, color='b')
ax.fill_between(iterations, q25, q75)
ax.set_xlabel('Iterations')
ax.set_ylabel('Normalized error')
ax.grid(True, which='both')
plt.show()
plt.savefig("qcqp_convergence.pdf")
